tests/appsec/waf/test_telemetry.py

```diff
@@ -1,28 +1,218 @@
+            self._assert_valid_tags(
+                full_tags=full_tags, valid_prefixes=valid_tag_prefixes, mandatory_prefixes=mandatory_tag_prefixes
+            )
+
+            if len(full_tags & {"request_blocked:false", "rule_triggered:false"}) == 2:
+                matched_not_blocked += 1
+                assert p[1] >= 1
+            elif len(full_tags & {"request_blocked:false", "rule_triggered:true"}) == 2:
+                matched_triggered += 1
+                assert p[1] == 1
+            elif len(full_tags & {"request_blocked:true", "rule_triggered:true"}) == 2:
+                matched_blocked += 1
+                assert p[1] == 1
+
+        assert matched_not_blocked >= 1
+        assert matched_triggered == 1
+        assert matched_blocked == 1
+
+    def _find_series(self, request_type, namespace, metric):
+        series = []
+        for data in interfaces.library.get_telemetry_data():
+            content = data["request"]["content"]
+            if content.get("request_type") != request_type:
+                continue
+            fallback_namespace = content["payload"].get("namespace")
+            for serie in content["payload"]["series"]:
+                computed_namespace = serie.get("namespace", fallback_namespace)
+                # Inject here the computed namespace considering the fallback. This simplifies later assertions.
+                serie["_computed_namespace"] = computed_namespace
+                if computed_namespace == namespace and serie["metric"] == metric:
+                    series.append(serie)
+        return series
+
+    def _assert_valid_tags(self, full_tags, valid_prefixes, mandatory_prefixes):
+        full_tags = set(full_tags)
+        tag_prefixes = {t.split(":")[0] for t in full_tags}
+
+        invalid_tags = tag_prefixes - valid_prefixes
+        assert not invalid_tags
+
+        missing_tags = mandatory_prefixes - tag_prefixes
+        assert not missing_tags
+
+
+def _validate_headers(headers, request_type):
+    """https://github.com/DataDog/instrumentation-telemetry-api-docs/blob/main/GeneratedDocumentation/ApiDocs/v2/how-to-use.md"""
+
+    expected_language = context.library.library
+    if expected_language == "java":
+        expected_language = "jvm"
+
     # empty value means we don't care about the content, but we want to check the key exists
+    # a set means "any of"
     expected_headers = {
-        "Content-Type": "application/json",
+        "Content-Type": {"application/json", "application/json; charset=utf-8"},
         "DD-Telemetry-API-Version": "v1",
         "DD-Telemetry-Request-Type": request_type,
-        "DD-Client-Library-Language": context.library,
+        "DD-Client-Library-Language": expected_language,
         "DD-Client-Library-Version": "",
-        "DD-Agent-Env": "system-tests",
-        "DD-Agent-Hostname": "",
     }
 
     for key, value in headers:
         if expected_headers.get(key) is not None:
```

Comment:
Do we have to match the header name casing? lower-case header names won't validate...